spec/02-integration/05-proxy/05-ssl_spec.lua

```diff
@@ -147,6 +170,59 @@ describe("SSL", function()
     end)
   end)
 
+  describe("proxy_ssl_name", function()
+    local https_client_sni
+
+    before_each(function()
+      assert(helpers.kong_exec("restart --conf " .. helpers.test_conf_path ..
+                               " --nginx-conf spec/fixtures/custom_nginx.template"))
+
+      https_client_sni = helpers.proxy_ssl_client()
+    end)
+
+    after_each(function()
+      https_client_sni:close()
+    end)
+
+    describe("properly sets the upstream SNI with preserve_host", function()
+      it("true", function()
+        local res = assert(https_client_sni:send {
+          method = "GET",
+          path = "/ssl-inspect",
+          headers = {
+            Host = "ssl3.com"
+          }
+        })
+        local body = assert.res_status(200, res)
+        assert.equal("ssl3.com", body)
+      end)
+
+      it("false", function()
+        local res = assert(https_client_sni:send {
+          method = "GET",
+          path = "/ssl-inspect",
+          headers = {
+            Host = "no-sni.com"
+          }
+        })
+        local body = assert.res_status(200, res)
+        assert.equal("localhost", body)
+      end)
+
+      it("false and IP-based upstream_url", function()
+        local res = assert(https_client_sni:send {
+          method = "GET",
+          path = "/ssl-inspect",
+          headers = {
+            Host = "nil-sni.com"
+          }
+        })
+        local body = assert.res_status(200, res)
+        assert.equal("nil", body)
```

Comment:
I think this test could become a false positive if the `nil` value is set from, say, a `tostring()` call on `upstream_host` when setting it with a nil value. Which could send a Client Hello with `nil`, and is thus given back by the fixture plugin.

Instead, I suggest that the fixture plugin tests for actual `nil` values, and sends back a response like `"no SNI"`, or an empty body even, to be more explicit and avoid such a false positive and where the `nil` is coming from.